### PR TITLE
CEC-1214:ServiceNow Connection Management Implementation

### PR DESCRIPTION
**Pull request by Shubhangi-cs**

### CODE REVIEW

File: src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowTableAPIClientImpl.java [hunk 1]
```diff
@@ -59,15 +59,20 @@ public class ServiceNowTableAPIClientImpl extends RestAPIClient {
   private static final String FIELD_CREATED_ON = "sys_created_on";
   private static final String FIELD_UPDATED_ON = "sys_updated_on";
   private static final String OAUTH_URL_TEMPLATE = "%s/oauth_token.do";
-  private ServiceNowBaseConfig conf;
-  
-  public ServiceNowTableAPIClientImpl(ServiceNowBaseConfig conf) {
+  private final ServiceNowConnectorConfig conf;
+  //private ServiceNowConnectorConfig connectorConfig;
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowTableAPIClientImpl.java]: Remove commented code from all files.

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+import java.util.List;
+import java.util.Map;
+import javax.ws.rs.core.MediaType;
+
+
+/**
+ * ServiceNow Connector Plugin
+ */
+@Plugin(type = Connector.PLUGIN_TYPE)
+@Name(ServiceNowConnector.NAME)
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: Use ServiceNowConstants.PLUGIN_NAME

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        canBrowse(false).canSample(true));
+      entity.addProperty(LABEL_NAME, BrowseEntityPropertyValue.builder(label, BrowseEntityPropertyValue.
+        PropertyType.STRING).build());
+      browseDetailBuilder.addEntity(entity.build());
+      count++;
+    }
+    return browseDetailBuilder.setTotalCount(count).build();
+  }
+
+  public TableList listTables() {
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: Add comments on top of the method. Make it private if not using elsewhere.

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    ServiceNowTableAPIRequestBuilder requestBuilder = new ServiceNowTableAPIRequestBuilder(
+      config.getRestApiEndpoint(), OBJECT_TABLE_LIST);
+    String accessToken = null;
+    ServiceNowTableAPIClientImpl serviceNowTableAPIClient = new ServiceNowTableAPIClientImpl(config);
+    try {
+      accessToken = serviceNowTableAPIClient.getAccessToken();
+    } catch (OAuthSystemException | OAuthProblemException e) {
+      e.printStackTrace();
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: remove e.printStackTrace and wrap exceptions in IOException and throw.

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+  @Override
+  public BrowseDetail browse(ConnectorContext connectorContext, BrowseRequest browseRequest) throws IOException {
+    int count = 0;
+    FailureCollector collector = connectorContext.getFailureCollector();
+    config.validateCredentials(collector);
+    collector.getOrThrowException();
+    BrowseDetail.Builder browseDetailBuilder = BrowseDetail.builder();
+    TableList tables = listTables();
+    Table[] table = tables.getResult();
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: directly use listTables.getResult

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+  @Override
+  public void test(ConnectorContext connectorContext) throws ValidationException {
+    FailureCollector collector = connectorContext.getFailureCollector();
+    config.validateCredentials(collector);
+  }
+
+  @Override
+  public BrowseDetail browse(ConnectorContext connectorContext, BrowseRequest browseRequest) throws IOException {
+    int count = 0;
+    FailureCollector collector = connectorContext.getFailureCollector();
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java:88]: remove collector.

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+      e.printStackTrace();
+    }
+    requestBuilder.setAuthHeader(accessToken);
+    requestBuilder.setAcceptHeader(MediaType.APPLICATION_JSON);
+    requestBuilder.setContentTypeHeader(MediaType.APPLICATION_JSON);
+    RestAPIResponse apiResponse = null;
+    apiResponse = serviceNowTableAPIClient.executeGet(requestBuilder.build());
+    if (!apiResponse.isSuccess()) {
+      LOG.error("Error - {}", getErrorMessage(apiResponse.getResponseBody()));
+      throw new RuntimeException();
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: throw IOException

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowInputFormat.java [hunk 8]
```diff
@@ -61,27 +64,31 @@ public static List<ServiceNowTableInfo> setInput(Configuration jobConfig, Source
...
 
     jobConf.setTableInfos(tableInfos);
 
     return tableInfos;
   }
 
-  static List<ServiceNowTableInfo> fetchTableInfo(SourceQueryMode mode, ServiceNowSourceConfig conf) {
+  public static List<ServiceNowTableInfo> fetchTableInfo(SourceQueryMode mode, ServiceNowConnectorConfig conf,
+                                                         String tableName, SourceApplication application,
+                                                         SourceValueType valueType, String startDate, String endDate) {
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/servicenow/source/ServiceNowInputFormat.java]: put @Nullable for all the parameters which can be null.

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+                                                       properties)).
+      addRelatedPlugin(new PluginSpec(ServiceNowConstants.PLUGIN_NAME, BatchSink.PLUGIN_TYPE, properties)).build();
+  }
+
+
+  private String getErrorMessage(String responseBody) {
+    try {
+      Gson gson = new Gson();
+      JsonObject jo = gson.fromJson(responseBody, JsonObject.class);
+      return jo.getAsJsonObject("error").get("message").getAsString();
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: Use constants in place of error and message. Sanchit must have added these in latest build.

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorConfig.java [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+  public String getUser() {
+    return user;
+  }
+
+  public String getPassword() {
+    return password;
+  }
+
+
+  public void validateCredentials(FailureCollector collector) {
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorConfig.java]: add documentation for all public methods.

Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorConfig.java]: similar method is also available in other config also. Reuse this one.

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowMultiInputFormat.java [hunk 11]
```diff
@@ -64,20 +67,23 @@ public static Set<ServiceNowTableInfo> setInput(Configuration jobConfig,
...
+                                                          conf.getValueType(),
+                                                          conf.getStartDate(), conf.getEndDate());
 
     jobConf.setTableInfos(tableInfos.stream().collect(Collectors.toList()));
 
     return tableInfos;
   }
 
-  static Set<ServiceNowTableInfo> fetchTablesInfo(ServiceNowMultiSourceConfig conf) {
+  static Set<ServiceNowTableInfo> fetchTablesInfo(ServiceNowConnectorConfig conf, String tableName,
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/servicenow/source/ServiceNowMultiInputFormat.java]: rename tableName to tableNames
set can be renamed to tableNameSet

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorConfig.java [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+
+
+/**
+ * PluginConfig for ServiceNow Connector
+ */
+public class ServiceNowConnectorConfig extends PluginConfig {
+
+  @Name(ServiceNowConstants.PROPERTY_CLIENT_ID)
+  @Macro
+  @Description(" The Client ID for ServiceNow Instance.")
```
Comment (sgarg-CS) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorConfig.java:33]: Remove extra space

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+@Plugin(type = Connector.PLUGIN_TYPE)
+@Name(ServiceNowConnector.NAME)
+@Description("Connection to access data in Servicenow tables.")
+public class ServiceNowConnector implements Connector {
+  public static final String NAME = "ServiceNow";
+  private static final Logger LOG = LoggerFactory.getLogger(ServiceNowConnector.class);
+  private static final String OBJECT_TABLE_LIST = "sys_db_object";
+  private static final String ENTITY_TYPE_TABLE = "table";
+  private static final String LABEL_NAME = "label";
+  private static final String OAUTH_URL_TEMPLATE = "%s/oauth_token.do";
```
Comment (sgarg-CS) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: Remove, if unused.

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+    requestBuilder.setContentTypeHeader(MediaType.APPLICATION_JSON);
+    RestAPIResponse apiResponse = null;
+    apiResponse = serviceNowTableAPIClient.executeGet(requestBuilder.build());
+    if (!apiResponse.isSuccess()) {
+      LOG.error("Error - {}", getErrorMessage(apiResponse.getResponseBody()));
+      throw new RuntimeException();
+    } else {
+      String response = null;
+      response = apiResponse.getResponseBody();
+      Gson gson = new Gson();
```
Comment (sgarg-CS) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: Define `Gson gson = new Gson();` at the class level.

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+    specBuilder.setSchema(schema);
+    return specBuilder.addRelatedPlugin(new PluginSpec(ServiceNowConstants.PLUGIN_NAME, BatchSource.PLUGIN_TYPE,
+                                                       properties)).
+      addRelatedPlugin(new PluginSpec(ServiceNowConstants.PLUGIN_NAME, BatchSink.PLUGIN_TYPE, properties)).build();
+  }
+
+
+  private String getErrorMessage(String responseBody) {
+    try {
+      Gson gson = new Gson();
```
Comment (sgarg-CS) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: Use the class level variable for `gson` after defining it.

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+      response = apiResponse.getResponseBody();
+      Gson gson = new Gson();
+      TableList tables = gson.fromJson(response, TableList.class);
+      return tables;
+    }
+  }
+
+  @Override
+  public ConnectorSpec generateSpec(ConnectorContext connectorContext, ConnectorSpecRequest connectorSpecRequest)
+    throws IOException {
```
Comment (sgarg-CS) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: When does this method throw `IOException`?

File: src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowTableAPIClientImpl.java [hunk 17]
```diff
@@ -59,15 +59,20 @@ public class ServiceNowTableAPIClientImpl extends RestAPIClient {
...
+
+  public ServiceNowTableAPIClientImpl(ServiceNowConnectorConfig conf) {
     this.conf = conf;
   }
-  
+  /*public ServiceNowTableAPIClientImpl(ServiceNowConnectorConfig connectorConfig) {
```
Comment (dhiraj0116) [inline src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowTableAPIClientImpl.java]: remove commented code

File: src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowTableAPIClientImpl.java [hunk 18]
```diff
@@ -139,7 +144,7 @@ public void createRecord(String tableName, HttpEntity entity) {
       if (!apiResponse.isSuccess()) {
         LOG.error("Error - {}", getErrorMessage(apiResponse.getResponseBody()));
       } else {
-        LOG.info(apiResponse.getResponseBody().toString());
+        LOG.info(apiResponse.getResponseBody());
```
Comment (dhiraj0116) [inline src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowTableAPIClientImpl.java:143]: Why do we need this?

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+    String tableName = connectorSpecRequest.getPath();
+    if (tableName != null) {
+      properties.put(ServiceNowConstants.PROPERTY_TABLE_NAME, tableName);
+    }
+    SourceQueryMode mode = SourceQueryMode.TABLE;
+    SourceValueType valueType = SourceValueType.SHOW_DISPLAY_VALUE;
+    List<ServiceNowTableInfo> tableInfo = ServiceNowInputFormat.fetchTableInfo(mode, config, tableName,
+                                                                               null, valueType,
+                                                                               null, null);
+    Schema schema = tableInfo.stream().findFirst().get().getSchema();
```
Comment (sgarg-CS) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: `tableInfo` can be an empty list, use the `Optional.isPresent()` check before calling get()

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java [hunk 20]
```diff
@@ -0,0 +1,167 @@
...
+    int count = 0;
+    FailureCollector collector = connectorContext.getFailureCollector();
+    config.validateCredentials(collector);
+    collector.getOrThrowException();
+    BrowseDetail.Builder browseDetailBuilder = BrowseDetail.builder();
+    TableList tables = listTables();
+    Table[] table = tables.getResult();
+    for (int i = 0; i < table.length; i++) {
+      String name = table[i].getName();
+      String label = table[i].getLabel();
```
Comment (dhiraj0116) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java:92]: We can remove name and label variable

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 14 above)
Comment (dhiraj0116) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: and make that private final

File: src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowTableAPIClientImpl.java (hunk 1 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowTableAPIClientImpl.java]: Removed.

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 2 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: Done

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 3 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: Done

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 4 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: We need these exceptions here for getAccessToken().

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 5 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: Done

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 7 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: Done

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowInputFormat.java (hunk 8 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/source/ServiceNowInputFormat.java]: Done

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 9 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: Done

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorConfig.java (hunk 10 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorConfig.java]: Done

Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorConfig.java]: It's better to use separate validation method here as the validation method which is defined in another config class includes the other validations as well which is not needed here.

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowMultiInputFormat.java (hunk 11 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/source/ServiceNowMultiInputFormat.java]: Done

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorConfig.java (hunk 12 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorConfig.java:33]: Done

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 13 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: Removed

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 14 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: Declaring it at class level is throwing NullPointer Exception.

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 15 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: Done

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 16 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: Removed

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 19 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: Done

File: src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowTableAPIClientImpl.java (hunk 17 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowTableAPIClientImpl.java]: Removed

File: src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowTableAPIClientImpl.java (hunk 18 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowTableAPIClientImpl.java:143]: This method is already written in the previous code, not modified for the present implementation. It's just that toString() has been removed as it was unused.

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 20 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java:92]: It is the standard implementation which has been done in other connectors as well. Do u want me to remove it?

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorConfig.java (hunk 10 above)
Comment (vikasrathee-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorConfig.java]: for similar validations you can call this method from that method and for rest of the validations keep that method as it is.

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 14 above)
Comment (vikasrathee-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: that should not happen, make sure you have not defined it elsewhere.

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 4 above)
Comment (vikasrathee-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: remove printstackTrace and write like
throw new IOException(e);

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java [hunk 21]
```diff
@@ -0,0 +1,167 @@
...
+    apiResponse = serviceNowTableAPIClient.executeGet(requestBuilder.build());
+    if (!apiResponse.isSuccess()) {
+      LOG.error("Error - {}", getErrorMessage(apiResponse.getResponseBody()));
+      throw new RuntimeException();
+    } else {
+      String response = null;
+      response = apiResponse.getResponseBody();
+      Gson gson = new Gson();
+      TableList tables = gson.fromJson(response, TableList.class);
+      return tables;
```
Comment (sgarg-CS) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: Can be simplified to `return gson.fromJson(response, TableList.class);`

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowSource.java [hunk 22]
```diff
@@ -58,11 +59,13 @@ public class ServiceNowSource extends BatchSource<NullWritable, StructuredRecord
   private static final Logger LOG = LoggerFactory.getLogger(ServiceNowSource.class);
 
   private final ServiceNowSourceConfig conf;
+  //private final ServiceNowConnectorConfig connectorConfig;
```
Comment (sgarg-CS) [inline src/main/java/io/cdap/plugin/servicenow/source/ServiceNowSource.java]: Remove commented code

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowRecordReader.java [hunk 23]
```diff
@@ -37,6 +37,7 @@
 public class ServiceNowRecordReader extends ServiceNowBaseRecordReader {
   private static final Logger LOG = LoggerFactory.getLogger(ServiceNowRecordReader.class);
   private final ServiceNowSourceConfig pluginConf;
+ // private ServiceNowConnectorConfig connectorConfig;
```
Comment (sgarg-CS) [inline src/main/java/io/cdap/plugin/servicenow/source/ServiceNowRecordReader.java]: Remove commented code

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowRecordReader.java [hunk 24]
```diff
@@ -20,10 +20,10 @@
 import io.cdap.cdap.api.data.schema.Schema;
 import io.cdap.plugin.servicenow.apiclient.ServiceNowTableAPIClientImpl;
 import io.cdap.plugin.servicenow.apiclient.ServiceNowTableDataResponse;
+import io.cdap.plugin.servicenow.connector.ServiceNowConnectorConfig;
```
Comment (sgarg-CS) [inline src/main/java/io/cdap/plugin/servicenow/source/ServiceNowRecordReader.java]: Remove unused import statements

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 4 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: Done

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorConfig.java (hunk 10 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorConfig.java]: Done

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 21 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: Done

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowSource.java (hunk 22 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/source/ServiceNowSource.java]: Done

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowRecordReader.java (hunk 23 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/source/ServiceNowRecordReader.java]: Done

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowRecordReader.java (hunk 24 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/source/ServiceNowRecordReader.java]: Done

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 14 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: Done.

**Pull request closed without merging**